File: exp-12/q1-bellman-ford/solution.py
import sys

INF = 999999


def main():
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    edges = []
    for _ in range(m):
        u = int(data[idx]); v = int(data[idx + 1]); w = int(data[idx + 2])
        idx += 3
        edges.append((u, v, w))
    s = int(data[idx])

    dist = [INF] * (n + 1)
    dist[s] = 0

    for _ in range(n - 1):
        changed = False
        for u, v, w in edges:
            if dist[u] != INF and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                changed = True
        if not changed:
            break

    for u, v, w in edges:
        if dist[u] != INF and dist[u] + w < dist[v]:
            print("Negative cycle detected")
            return

    print(f"Shortest distances from source {s}:")
    for i in range(1, n + 1):
        if dist[i] == INF:
            print(f"{i}: INF")
        else:
            print(f"{i}: {dist[i]}")


if __name__ == "__main__":
    main()

File: exp-12/q1-bellman-ford/solution.go
package main

import "fmt"

const INF = 999999

type Edge struct {
	u, v, w int
}

func main() {
	var n, m, s int
	fmt.Scan(&n, &m)

	edges := make([]Edge, m)
	for i := 0; i < m; i++ {
		fmt.Scan(&edges[i].u, &edges[i].v, &edges[i].w)
	}
	fmt.Scan(&s)

	dist := make([]int, n+1)
	for i := 1; i <= n; i++ {
		dist[i] = INF
	}
	dist[s] = 0

	for i := 1; i <= n-1; i++ {
		changed := false
		for _, e := range edges {
			if dist[e.u] != INF && dist[e.u]+e.w < dist[e.v] {
				dist[e.v] = dist[e.u] + e.w
				changed = true
			}
		}
		if !changed {
			break
		}
	}

	for _, e := range edges {
		if dist[e.u] != INF && dist[e.u]+e.w < dist[e.v] {
			fmt.Println("Negative cycle detected")
			return
		}
	}

	fmt.Printf("Shortest distances from source %d:\n", s)
	for i := 1; i <= n; i++ {
		if dist[i] == INF {
			fmt.Printf("%d: INF\n", i)
		} else {
			fmt.Printf("%d: %d\n", i, dist[i])
		}
	}
}
